src/lib.rs

```diff
@@ -884,6 +921,215 @@ pub unsafe trait FromBytes: FromZeroes {
+/// including changes that only affect soundness, which may cause code which
+/// uses those items to silently become unsound.
+///
+/// [`try_from_ref`]: TryFromBytes::try_from_ref
+pub unsafe trait TryFromBytes: KnownLayout {
+    /// Does a given memory range contain a valid instance of `Self`?
+    ///
+    /// # Safety
+    ///
+    /// ## Preconditions
+    ///
+    /// `candidate` must be a validly-aligned pointer, and it must be [valid]
+    /// for reads. The total length encoded by the pointer must not overflow an
+    /// `isize`. The memory addressed by `c` must fall within a single
+    /// allocation.
+    ///
+    /// `candidate`'s referent is not required to contain a valid `Self`.
+    /// However, it must satisfy the requirement that uninitialized bytes may
+    /// only be present where it is possible for them to be present in `Self`.
+    /// This is a dynamic property: if, at a particular byte offset, a valid
+    /// enum discriminant is set, the subsequent bytes may only have
+    /// uninitialized bytes as specificed by the corresponding enum.
+    ///
+    /// Formally, given `len = size_of_val_raw(candidate)`, at every byte
+    /// offset, `b`, in the range `[0, len)`:
+    /// - If, in all instances `s: Self` of length `len`, the byte at offset `b`
+    ///   in `s` is initialized, then the byte at offset `b` within `*candidate`
+    ///   must be initialized.
+    /// - Let `c` be the contents of the byte range `[0, b)` in `*candidate`.
+    ///   Let `S` be the subset of valid instances of `Self` of length `len`
+    ///   which contain `c` in the offset range `[0, b)`. If, for all instances
+    ///   of `s: Self` in `S`, the byte at offset `b` in `s` is initialized,
+    ///   then the byte at offset `b` in `*candidate` must be initialized.
+    ///
+    ///   Pragmatically, this means that if `*candidate` is guaranteed to
+    ///   contain an enum type at a particular offset, and the enum discriminant
+    ///   stored in `*candidate` corresponds to a valid variant of that enum
+    ///   type, then it is guaranteed that the appropriate bytes of `*candidate`
+    ///   are initialized as defined by that variant's bit validity (although
+    ///   note that the variant may contain another enum type, in which case the
+    ///   same rules apply depending on the state of its discriminant, and so on
+    ///   recursively).
+    ///
+    /// ## Postconditions
+    ///
+    /// Unsafe code may assume that, if `is_bit_valid(candidate)` returns true,
+    /// `*candidate` contains a valid `Self`.
+    ///
+    /// [valid]: https://doc.rust-lang.org/std/ptr/index.html#safety
+    #[doc(hidden)]
+    unsafe fn is_bit_valid(candidate: NonNull<Self>) -> bool;
+
+    /// Attempts to interpret a byte slice as a `Self`.
+    ///
+    /// `try_from_ref` validates that `bytes` contains a valid `Self` as defined
+    /// by [`is_bit_valid`]. If it does, then `bytes` is reinterpreted as a
+    /// `Self`.
+    ///
```

Comment:
TODO: Maybe document that we guarantee that our validity is a subset of Rust's? Ie, we might reject a type which Rust thinks is valid, but we'll never do the inverse.